File: extra-tests/cross_write.c
/* Write canonical MessagePack object to a file (C → Rust reader).
 * Usage: cross_write_c <out.mp>
 *
 * map(4):
 *   "compact" -> true
 *   "schema"  -> 0
 *   "name"    -> "demo"
 *   "vals"    -> array(3): -32, 255, nil
 */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#include "cwpack.h"

static int pack_object(uint8_t *buf, size_t buflen, size_t *out_len) {
    cw_pack_context pc;
    if (cw_pack_context_init(&pc, buf, (unsigned long)buflen, 0))
        return pc.return_code;

    cw_pack_map_size(&pc, 4);

    cw_pack_str(&pc, "compact", 7);
    cw_pack_boolean(&pc, true);

    cw_pack_str(&pc, "schema", 6);
    cw_pack_unsigned(&pc, 0);

    cw_pack_str(&pc, "name", 4);
    cw_pack_str(&pc, "demo", 4);

    cw_pack_str(&pc, "vals", 4);
    cw_pack_array_size(&pc, 3);
    cw_pack_signed(&pc, -32);
    cw_pack_unsigned(&pc, 255);
    cw_pack_nil(&pc);

    if (pc.return_code)
        return pc.return_code;
    *out_len = (size_t)(pc.current - pc.start);
    return 0;
}

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "usage: %s <out.mp>\n", argv[0]);
        return 2;
    }
    uint8_t buf[256];
    size_t n = 0;
    int rc = pack_object(buf, sizeof buf, &n);
    if (rc) {
        fprintf(stderr, "pack failed rc=%d\n", rc);
        return 3;
    }
    FILE *f = fopen(argv[1], "wb");
    if (!f) {
        perror("fopen");
        return 4;
    }
    if (fwrite(buf, 1, n, f) != n) {
        perror("fwrite");
        fclose(f);
        return 5;
    }
    fclose(f);
    fprintf(stderr, "c wrote %zu bytes -> %s\n", n, argv[1]);
    return 0;
}

File: examples/cross_write.rs
//! Write a fixed MessagePack "object" to a file (Rust → C reader).
//! Usage: cross_write <out.mp>

use cwpack::{
    cw_pack_array_size, cw_pack_boolean, cw_pack_map_size, cw_pack_nil, cw_pack_signed,
    cw_pack_str, cw_pack_unsigned, CwPackContext,
};
use std::env;
use std::fs;
use std::process::ExitCode;

/// Canonical object (same layout as extra-tests/cross_write.c):
/// map(4):
///   "compact" -> true
///   "schema"  -> 0
///   "name"    -> "demo"
///   "vals"    -> array(3): -32, 255, nil
fn pack_object(buf: &mut [u8]) -> Result<usize, i32> {
    let mut pc = CwPackContext::new(buf);
    cw_pack_map_size(&mut pc, 4);

    cw_pack_str(&mut pc, b"compact", 7);
    cw_pack_boolean(&mut pc, true);

    cw_pack_str(&mut pc, b"schema", 6);
    cw_pack_unsigned(&mut pc, 0);

    cw_pack_str(&mut pc, b"name", 4);
    cw_pack_str(&mut pc, b"demo", 4);

    cw_pack_str(&mut pc, b"vals", 4);
    cw_pack_array_size(&mut pc, 3);
    cw_pack_signed(&mut pc, -32);
    cw_pack_unsigned(&mut pc, 255);
    cw_pack_nil(&mut pc);

    if pc.return_code != 0 {
        return Err(pc.return_code);
    }
    Ok(pc.len_packed())
}

fn main() -> ExitCode {
    let path = env::args().nth(1).expect("usage: cross_write <out.mp>");
    let mut buf = [0u8; 256];
    let n = match pack_object(&mut buf) {
        Ok(n) => n,
        Err(rc) => {
            eprintln!("pack failed rc={rc}");
            return ExitCode::from(2);
        }
    };
    if let Err(e) = fs::write(&path, &buf[..n]) {
        eprintln!("write {path}: {e}");
        return ExitCode::from(3);
    }
    eprintln!("rust wrote {n} bytes -> {path}");
    ExitCode::SUCCESS
}
